### PR TITLE
VK/GL: Scale to game's aspect ratio before scaling to viewport

### LINKED ISSUE #8725: [Feature Request] Proper support for anamorphic video modes
480p and 576p are anamorphic video modes on a real PS3, they're used on SDTVs (CRTs) which rescale the image to fill the screen (which is usually 4:3 but there are some 16:9 SDTVs). RPCS3 simply displays these modes in pixel aspect ratio which is incorrect and makes the game appear distorted. You can get around this by forcing a specific window size and enabling "Stretch to Display Area" but this is far from optimal.

Example from SONIC THE HEDGEHOG 4 Episode I [NPEB00153] (running in 480p 4:3):
image

Correct aspect ratio:
image

Same game running in 720p:
image

### LINKED ISSUE #11099: God of War Collection [BCES00791]: Narrow graphics at 1080p
Both games render incorrectly when the default resolution is 1080p.

God of War 1 HD:
Error1
[RPCS3 (1).log](https://github.com/RPCS3/rpcs3/files/7459882/RPCS3.1.log)

God of War 2 HD:
Error2
[RPCS3 (2).log](https://github.com/RPCS3/rpcs3/files/7459883/RPCS3.2.log)

### LINKED ISSUE #11143: RSX: Some (or maybe many?) members of rsx::thread and its subclasses aren't properly reset between boots
Contrary to what you'd expect, rsx threads aren't actually destroyed between boots, which means they have to be manually initialized and reset when the emulator stops.

This small detail can lead to human error, which means developers may not know this and simply add a bunch of members, believing that the stuff is surely reset on the next boot, or they may just forget to deal with this properly.

Why do I tell you this? Because currently there are a bunch of members that are **not** properly reset and therefore contain old data that is carried over to the next game boot.

One such example would be `display_buffers`, which starts as a zero initialized array of objects on the first boot, but magically contains random data on the second game boot.

Plz fix, thx

### PR DESCRIPTION
**Pull request by Megamouse**
I noticed that we never ever look for the game's aspect ratio when we scale the image to our viewport.

fixes #11099
fixes #8725
fixes #11143

### CODE REVIEW

**Comment by capriots** [comment]
Works great on my end, fixes #11099 and also #8725

**Comment by Megamouse** [comment]
Added screenshot scaling as well

File: rpcs3/Emu/RSX/rsx_utils.cpp [hunk 1]
```diff
@@ -102,6 +103,94 @@ namespace rsx
...
+			height = static_cast<int>(height / scaling_factor);
+		}
+	}
+
+	void avconf::downscale_to_aspect_ratio(int& x, int& y, int& width, int& height) const
+	{
+		if (width == 0 || height == 0) return;
+
+		const double old_aspect = 1. * width / height;
+		const double scaling_factor = get_aspect_ratio() / old_aspect;
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_utils.cpp:178]: The numerator should be (buffer_width/buffer_height) and not get_aspect_ratio().

**Comment by kd-11** [comment]
I'm working on fixing the letterboxing issue.

**Comment by kd-11** [comment]
External PR: https://github.com/Megamouse/rpcs3/pull/8

**Pull request merged**